File: sections/63_atcoder_difficulty_ladder/problems/42_random_is/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    move_count, pile_count = data[0], data[1]
    moves = data[2 : 2 + move_count]
    piles = data[2 + move_count : 2 + move_count + pile_count]
    maximum = max(piles)
    grundy = [0] * (maximum + 1)
    for stones in range(1, maximum + 1):
        reachable = {grundy[stones - move] for move in moves if move <= stones}
        while grundy[stones] in reachable:
            grundy[stones] += 1
    total = 0
    for stones in piles:
        total ^= grundy[stones]
    print("WIN" if total else "LOSE")


if __name__ == "__main__":
    main()

File: sections/63_atcoder_difficulty_ladder/problems/42_random_is/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int move_count, pile_count;
    if (!(cin >> move_count >> pile_count)) return 0;
    vector<int> moves(move_count), piles(pile_count);
    for (int& move : moves) cin >> move;
    int maximum = 0;
    for (int& pile : piles) {
        cin >> pile;
        maximum = max(maximum, pile);
    }

    vector<int> grundy(maximum + 1);
    for (int stones = 1; stones <= maximum; ++stones) {
        bool seen[32]{};
        for (int move : moves) {
            if (move <= stones) seen[grundy[stones - move]] = true;
        }
        while (seen[grundy[stones]]) ++grundy[stones];
    }
    int total = 0;
    for (int pile : piles) total ^= grundy[pile];
    cout << (total ? "WIN" : "LOSE") << '\n';
}
